### PR TITLE
Mindist for constraints

### PR DESCRIPTION
**Pull request by davidsean**
Fixes #

Description of changes:
 - wrote a core mindist for constraints
 - python interface
 - updated documentation
 - fixed a warmup problem for the constraints sample
 - CI tests for wall and cylinder

PR Checklist
------------
 - [x] Tests?
   - [x] Interface
   - [x] Core 
 - [x] Docs?

### CODE REVIEW

File: testsuite/constraint_shape_based_wall.py [hunk 1]
```diff
@@ -45,14 +45,15 @@ def test(self):
         # (1)
         constraint_wy = espressomd.constraints.ShapeBasedConstraint(
             shape=wy, particle_type=1)
-        wall_xz = system.constraints.add(constraint_wy)
+        wall_xy = system.constraints.add(constraint_wy)
```
Comment (jonaslandsgesell) [inline testsuite/constraint_shape_based_wall.py]: Why did you exchange the naming of the wall variables? The old convention was the following: if the normal vector is in y direction (0,1,0) then the wall lies parallel to the xz plane and therefore the wall is named wall_xz

File: src/core/constraints/ShapeBasedConstraint.cpp [hunk 2]
```diff
@@ -17,6 +18,30 @@ Vector3d ShapeBasedConstraint::total_force() const {
     return total_force;
 }
 
+
+double ShapeBasedConstraint::mindist() {
```
Comment (fweik) [inline src/core/constraints/ShapeBasedConstraint.cpp]: This is parallel code (it runs on all nodes in synchronous mode), so you can't use `partCfg`, as it exists only on the head node. There is also no need, as your algorithm does not need global information, you can run it on each node locally and then reduce the local results to the master. Also folded coordinates should be used with the shape, but particles in partCfg have folded coordinates. So you could do something along the lines of

```c++
auto & parts = local_cells.particles();

auto const local_mindist = std::accumulate(parts.begin(), parts.end(), std::numeric_limits<double>::infinity(), [this](double min, Particle const& p) {
  if(!checkIfInteraction(...))
    return min;
  else  {
     double dummy[3], dist;
     m_shape->calculate_dist(folded_position(p).data(), &dist, dummy); 

     return std::min(min, dist);
   });

   double global_min;
   boost::mpi::reduce(comm_cart, local_min, global_min, std::min, 0);

   return global_min;
}
```  
}

**Review by fweik — CHANGES_REQUESTED** [review]
There are several issues with this:
- It uses the shape contained in the constraint with unfolded coordinates where the shape expects folded positions. (positions in partCfg are unfolded)
- The tests apparently did not catch this.
- It uses partCfg that can only be used on the head node on all nodes.

**Comment by RudolfWeeber** [comment]
The python method in analysis.pyx is called min_dist.
I'm also not sure, min_dist is actually what it does. For a sphere,
e.g., I think it takes the vector which is perpendicular to the
surface of the sphere (@pkreissl?).

File: testsuite/constraint_shape_based_wall.py (hunk 1 above)
Comment (davidsean) [reply testsuite/constraint_shape_based_wall.py]: I had it follow the naming convention for walls and shape where the last character gives the normal. I did not realize xz meant the xz plane. I don't care which convention is chosen, however constraint_wy should then be constraint_xz and the shape wy should be xz to be consistent.

**Review by fweik — APPROVED** [review]

**Pull request merged**